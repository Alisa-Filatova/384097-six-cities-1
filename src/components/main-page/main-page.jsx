import React from 'react';
import PropTypes from 'prop-types';
import OffersList from '../offers-list/offers-list.jsx';
import Map from '../map/map.jsx';
import TownsList from '../towns-list/towns-list.jsx';
import withActiveItem from '../../hocs/with-active-item.jsx';
import withTransformProps from '../../hocs/with-transform-props.jsx';

const WrappedOffersList = withActiveItem(
    withTransformProps((props) => Object.assign({}, props, {
      setActiveItem: props.setActiveItem,
    }))(OffersList)
);

class MainPage extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      activeOfferId: null,
    };
  }

  render() {
    const {rentalOffers, onTownClick, currentTown, cityOffers} = this.props;

    return (
      <main className="page__main page__main--index">
        <h1 className="visually-hidden">Cities</h1>
        <TownsList
          offers={rentalOffers}
          currentTown={currentTown}
          onTownClick={onTownClick}
        />
        <div className="cities__places-wrapper" style={{height: `100vh`}}>
          <div className="cities__places-container container">
            <section className="cities__places places">
              <h2 className="visually-hidden">Places</h2>
              <b className="places__found">
                {`${cityOffers.length} ${cityOffers.length === 1 ? `place` : `places`} to stay in ${currentTown.name}`}
              </b>
              <form className="places__sorting" action="#" method="get">
                <span className="places__sorting-caption">Sort by</span>
                <span className="places__sorting-type" tabIndex="0">
                  Popular
                  <svg className="places__sorting-arrow" width="7" height="4">
                    <use xlinkHref="#icon-arrow-select"/>
                  </svg>
                </span>
                <ul className="places__options places__options--custom places__options--opened">
                  <li className="places__option places__option--active" tabIndex="0">Popular</li>
                  <li className="places__option" tabIndex="0">Price: low to high</li>
                  <li className="places__option" tabIndex="0">Price: high to low</li>
                  <li className="places__option" tabIndex="0">Top rated first</li>
                </ul>
              </form>
              <WrappedOffersList
                rentalOffers={cityOffers}
                setActiveItem={this._handleGetActiveOffer.bind(this)}
              />
            </section>
            <div className="cities__right-section">
              <Map
                key={currentTown.name}
                currentTown={currentTown}
                activeOfferId={this.state.activeOfferId}
                cityOffers={cityOffers}
              />
            </div>
          </div>
        </div>
      </main>
    );
  }

  _handleGetActiveOffer(offerId) {
    this.setState((prevState) => {
      return Object.assign({}, prevState, {activeOfferId: offerId});
    });
  }
}

MainPage.propTypes = {
  rentalOffers: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number,
    title: PropTypes.string,
    [`preview_image`]: PropTypes.string,
    images: PropTypes.arrayOf(PropTypes.string),
    [`is_premium`]: PropTypes.bool,
    [`is_favorite`]: PropTypes.bool,
    bedrooms: PropTypes.number,
    goods: PropTypes.arrayOf(PropTypes.string),
    description: PropTypes.string,
    price: PropTypes.number,
    rating: PropTypes.number,
    type: PropTypes.string,
    location: PropTypes.shape({
      latitude: PropTypes.number,
      longitude: PropTypes.number,
      zoom: PropTypes.number,
    }),
    city: PropTypes.shape({
      name: PropTypes.string,
      location: PropTypes.shape({
        latitude: PropTypes.number,
        longitude: PropTypes.number,
        zoom: PropTypes.number,
      }),
    }),
    host: PropTypes.shape({
      id: PropTypes.number,
      name: PropTypes.string,
      [`is_pro`]: PropTypes.bool,
      [`avatar_url`]: PropTypes.string,
    }),
  })).isRequired,
  currentTown: PropTypes.object.isRequired,
  onTownClick: PropTypes.func.isRequired,
  cityOffers: PropTypes.array.isRequired,
};

export default MainPage;
