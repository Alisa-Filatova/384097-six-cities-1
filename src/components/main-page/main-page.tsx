import * as React from 'react';
import {connect} from 'react-redux';
import {ActionCreator, Operation} from '../../reducers/data/data';
import {
  getCurrentCity,
  getCities,
  getOffersLoadStatus,
  getSortValue,
  sortOffers, getOfferById,
} from '../../reducers/data/selectors';
import OffersList from '../offers-list/offers-list.jsx';
import Map from '../map/map.jsx';
import CitiesList from '../cities-list/cities-list.jsx';
import SortBy from '../sort-by/sort-by.jsx';
import MainPageEmpty from '../main-page-empty/main-page-empty.jsx';
import Loader from '../loader/loader.jsx';
import SortType from '../../types/enums/sort-type';
import {getAuthorizationStatus} from '../../reducers/user/selectors';
import withActiveOfferId from '../../hocs/with-active-offer-id/with-active-offer-id.jsx';
import withTransformProps from '../../hocs/with-transform-props/with-transform-props.jsx';
import {City, Offer} from '../../types/offer';
import {History} from '../../types/location';

interface Props {
  cities: City[];
  currentCity: City;
  onCityClick: () => void;
  cityOffers: Offer[];
  onPopularClick: () => void;
  onLowToHighClick: () => void;
  onHighToLowClick: () => void;
  onTopRatedClick: () => void;
  offersLoaded: boolean;
  sortValue: string;
  history: History,
  onFavoriteClick: (Offer) => void;
  isAuthenticated: boolean,
  offer: Offer;
  activeOfferId: number;
  onImgClick: () => void;
}

const MainPage: React.FunctionComponent<Props> = (props) => {
  const {
    onCityClick,
    currentCity,
    cityOffers,
    cities,
    offersLoaded,
    onPopularClick,
    onLowToHighClick,
    onHighToLowClick,
    onTopRatedClick,
    onFavoriteClick,
    sortValue,
    isAuthenticated,
    history,
    onImgClick,
    activeOfferId,
  } = props;

  return (
    <>
      <main className={`page__main page__main--index ${cityOffers.length === 0 ? `page__main--index-empty` : ``}`}>
        {!offersLoaded && <Loader />}
        {offersLoaded && cityOffers.length > 0 && (
          <>
            <h1 className="visually-hidden">Cities</h1>
            <CitiesList
              cities={cities}
              currentCity={currentCity}
              onCityClick={onCityClick}
            />
            <div className="cities__places-wrapper" style={{height: `100vh`}}>
              <div className="cities__places-container container">
                <section className="cities__places places">
                  <h2 className="visually-hidden">Places</h2>
                  <b className="places__found">
                    {`${cityOffers.length} ${cityOffers.length === 1 ? `place` : `places`} to stay in ${currentCity.name}`}
                  </b>
                  <SortBy
                    currentItem={sortValue}
                    onPopularClick={onPopularClick}
                    onLowToHighClick={onLowToHighClick}
                    onHighToLowClick={onHighToLowClick}
                    onTopRatedClick={onTopRatedClick}
                  />
                  <OffersList
                    rentalOffers={cityOffers}
                    onImgClick={onImgClick}
                    history={history}
                    onFavoriteClick={onFavoriteClick}
                    isAuthenticated={isAuthenticated}
                  />
                </section>
                <div className="cities__right-section">
                  <Map
                    key={currentCity.name}
                    currentCity={currentCity}
                    activeOfferId={activeOfferId}
                    cityOffers={cityOffers}
                    zoom
                  />
                </div>
              </div>
            </div>
          </>
        )}
        {offersLoaded && cityOffers.length === 0 && (
          <MainPageEmpty currentCity={currentCity} />
        )}
      </main>
    </>
  );
};

const mapStateToProps = (state, ownProps) => Object.assign({}, ownProps, {
  currentCity: getCurrentCity(state),
  cityOffers: sortOffers(state),
  cities: getCities(state),
  offersLoaded: getOffersLoadStatus(state),
  sortValue: getSortValue(state),
  offer: getOfferById(state, ownProps.activeOfferId),
  isAuthenticated: getAuthorizationStatus(state),
});

const mapDispatchToProps = (dispatch) => ({
  onCityClick: (currentCity) => {
    dispatch(ActionCreator.changeCity(currentCity));
  },
  onPopularClick: () => {
    dispatch(ActionCreator.sortOffers(SortType.POPULAR));
  },
  onLowToHighClick: () => {
    dispatch(ActionCreator.sortOffers(SortType.LOW_TO_HIGH));
  },
  onHighToLowClick: () => {
    dispatch(ActionCreator.sortOffers(SortType.HIGH_TO_LOW));
  },
  onTopRatedClick: () => {
    dispatch(ActionCreator.sortOffers(SortType.TOP_RATED));
  },
  onFavoriteClick: (offer) => {
    dispatch(Operation.changeFavorites(offer));
  },
});

export {MainPage};
export default withActiveOfferId(
    withTransformProps((props) => Object.assign({}, props, {
      onImgClick: props.setActiveId,
    })
    )(connect(mapStateToProps, mapDispatchToProps)(MainPage))
);
