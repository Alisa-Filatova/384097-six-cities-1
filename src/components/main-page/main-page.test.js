import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import {configure, shallow} from 'enzyme';
import {shallowToJson} from 'enzyme-to-json';
import {MainPage} from './main-page.jsx';

configure({adapter: new Adapter()});

const mock = [
  {
    id: 1,
    city: {
      name: `Amsterdam`,
      location: {
        latitude: 52.370216,
        longitude: 4.895168,
        zoom: 10,
      }
    },
    previewImage: `img/1.png`,
    images: [`img/1.png`, `img/2.png`],
    title: `Beautiful & luxurious studio at great location`,
    isFavorite: false,
    isPremium: false,
    rating: 4.8,
    type: `apartment`,
    bedrooms: 3,
    maxAdults: 4,
    price: 120,
    goods: [`Heating`, `Kitchen`, `Cable TV`, `Washing machine`, `Coffee machine`, `Dishwasher`],
    host: {
      id: 3,
      isPro: true,
      name: `Angelina`,
      avatarUrl: `img/1.png`
    },
    description: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.`,
    location: {
      latitude: 52.35514938496378,
      longitude: 4.673877537499948,
      zoom: 8,
    }
  }
];

describe(`MainPage`, () => {
  it(`renders correctly`, () => {
    const page = shallow(
        <MainPage
          cityOffers={mock}
          currentCity={mock[0].city}
          onCityClick={jest.fn()}
          cities={mock}
          offersLoaded={false}
          onPopularClick={jest.fn()}
          onLowToHighClick={jest.fn()}
          onHighToLowClick={jest.fn()}
          onTopRatedClick={jest.fn()}
          sortValue={`Popular`}
          history={{}}
        />
    );

    expect(shallowToJson(page)).toMatchSnapshot();
  });
});
