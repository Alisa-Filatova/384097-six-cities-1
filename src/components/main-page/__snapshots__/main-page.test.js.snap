// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage renders correctly 1`] = `
<main
  className="page__main page__main--index"
>
  <h1
    className="visually-hidden"
  >
    Cities
  </h1>
  <TownsList
    currentTown={
      Object {
        "location": 123,
      }
    }
    offers={
      Array [
        Object {
          "city": Object {
            "coordinates": Array [
              52.38333,
              4.9,
            ],
            "name": "Amsterdam",
          },
          "coordinates": Array [
            52.3909553943508,
            4.929309666406198,
          ],
          "id": 123,
          "isInBookmarks": false,
          "isPremium": true,
          "preview_image": "img/apartment-01.jpg",
          "price": 120,
          "rating": 4,
          "title": "Beautiful & luxurious apartment at great location",
          "type": "Apartment",
        },
        Object {
          "city": Object {
            "coordinates": Array [
              52.38333,
              4.9,
            ],
            "name": "Amsterdam",
          },
          "coordinates": Array [
            52.3909553943508,
            4.929309666406198,
          ],
          "id": 123,
          "isInBookmarks": false,
          "isPremium": true,
          "preview_image": "img/apartment-01.jpg",
          "price": 120,
          "rating": 4,
          "title": "Beautiful & luxurious apartment at great location",
          "type": "Apartment",
        },
      ]
    }
    onTownClick={[MockFunction]}
  />
  <div
    className="cities__places-wrapper"
    style={
      Object {
        "height": "100vh",
      }
    }
  >
    <div
      className="cities__places-container container"
    >
      <section
        className="cities__places places"
      >
        <h2
          className="visually-hidden"
        >
          Places
        </h2>
        <b
          className="places__found"
        >
          2 places to stay in undefined
        </b>
        <form
          action="#"
          className="places__sorting"
          method="get"
        >
          <span
            className="places__sorting-caption"
          >
            Sort by
          </span>
          <span
            className="places__sorting-type"
            tabIndex="0"
          >
            Popular
            <svg
              className="places__sorting-arrow"
              height="4"
              width="7"
            >
              <use
                xlinkHref="#icon-arrow-select"
              />
            </svg>
          </span>
          <ul
            className="places__options places__options--custom places__options--opened"
          >
            <li
              className="places__option places__option--active"
              tabIndex="0"
            >
              Popular
            </li>
            <li
              className="places__option"
              tabIndex="0"
            >
              Price: low to high
            </li>
            <li
              className="places__option"
              tabIndex="0"
            >
              Price: high to low
            </li>
            <li
              className="places__option"
              tabIndex="0"
            >
              Top rated first
            </li>
          </ul>
        </form>
        <WithActiveItem
          rentalOffers={
            Array [
              "Amsterdam",
              "Paris",
            ]
          }
          setActiveItem={[Function]}
        />
      </section>
      <div
        className="cities__right-section"
      >
        <Map
          activeOfferId={null}
          cityOffers={
            Array [
              "Amsterdam",
              "Paris",
            ]
          }
          currentTown={
            Object {
              "location": 123,
            }
          }
        />
      </div>
    </div>
  </div>
</main>
`;
